src/v1/index.ts

```diff
@@ -328,3 +341,114 @@ function decodeBallotVotes(contests: Contests, bits: BitReader): VotesDict {
 
   return votes
 }
+
+interface HMPBBallotPageMetadata {
+  election: Election
+  electionHash: string // a hexadecimal string
+  precinctId: string
+  ballotStyleId: string
+  locales: BallotLocale
+  pageNum: number
+  isLiveMode: boolean
+  isAbsenteeMode: boolean
+  ballotId?: string
+}
+
+export function encodeHMPBBallotPageMetadata({
+  election,
+  electionHash,
+  precinctId,
+  ballotStyleId,
+  locales,
+  pageNum,
+  isLiveMode,
+  isAbsenteeMode,
+  ballotId,
+}: HMPBBallotPageMetadata): Uint8Array {
+  const bits = new BitWriter()
+
+  const precincts = election.precincts
+  const ballotStyles = election.ballotStyles
+
+  const precinctIndex = precincts.findIndex((p) => p.id === precinctId)
+  const ballotStyleIndex = ballotStyles.findIndex(
+    (bs) => bs.id === ballotStyleId
+  )
+  const primaryLocaleIndex = SUPPORTED_LOCALES.indexOf(locales.primary)
+  const secondaryLocaleIndex = locales.secondary
+    ? SUPPORTED_LOCALES.indexOf(locales.secondary)
+    : undefined
+
+  bits
+    .writeUint8(...HMPBPrelude)
+    .writeString(electionHash, { encoding: HexEncoding })
+    .writeUint(precinctIndex, { max: precincts.length })
+    .writeUint(ballotStyleIndex, { max: ballotStyles.length })
+    .writeUint(primaryLocaleIndex, { max: SUPPORTED_LOCALES.length })
+    .writeBoolean(!!secondaryLocaleIndex)
+
+  if (secondaryLocaleIndex) {
+    bits.writeUint(secondaryLocaleIndex, { max: SUPPORTED_LOCALES.length })
+  }
+
+  bits.writeUint(pageNum, { max: 30 })
+
+  bits
+    .writeBoolean(isLiveMode)
+    .writeBoolean(isAbsenteeMode)
+    .writeBoolean(!!ballotId)
+
+  if (ballotId) {
+    bits.writeString(ballotId)
+  }
+
+  return bits.toUint8Array()
+}
+
+export function decodeHMPBBallotPageMetadata({
+  election,
+  data,
+}: {
+  election: Election
+  data: Uint8Array
+}): HMPBBallotPageMetadata {
+  const bits = new BitReader(data)
+
+  if (!bits.skipUint8(...HMPBPrelude)) {
+    throw new Error(
+      "expected leading prelude 'V' 'P' 0b00000001 but it was not found"
+    )
+  }
+
+  const precincts = election.precincts
+  const ballotStyles = election.ballotStyles
+
+  const electionHash = bits.readString({ encoding: HexEncoding })
```

Comment:
Anything after this point could fail/crash if the hash is not as expected. Worth separating this somehow so an explicit message can be shown comparing the hashes?

Reply:
I want to explicitly not crash / throw at this low level if there was a small text change that doesn't actually change the ballot layout, etc. I can imagine this happening by mistake. I'm thinking the hash comparison can happen at a higher level and we can choose to ignore it if we want to, enforce it if we prefer

Reply:
What I mean is that the `read*` calls below may fail given a different enough election definition. If there is 1 precinct and 1 ballot style in one election, then 90 of both in another, that second one will use more bits to represent the precinct and ballot style indexes. Every call after that would be reading with an incorrect offset. If the padding or offset works out unfavorably, the final `readString` call in this function could trigger an EOF error.

Reply:
Right. But then are you proposing that we hard fail if there is an election hash mismatch?

Reply:
No, I'm suggesting it _will_ fail as written in some circumstances. Thus, if we want a reliable way to read the hash for comparison, we either need to separate reading the hash from reading the rest, or fix the size of the precinct and ballot style uints.